src/engine/Service.cpp

```diff
@@ -0,0 +1,175 @@
+// matter and it should certainly not cause segmentation faults.
+#include "engine/Service.h"
+
+#include <absl/strings/str_cat.h>
+#include <absl/strings/str_split.h>
+
+#include "engine/CallFixedSize.h"
+#include "engine/Values.h"
+#include "engine/VariableToColumnMap.h"
+#include "parser/TokenizerCtre.h"
+#include "parser/TurtleParser.h"
+#include "util/Exception.h"
+#include "util/HashSet.h"
+#include "util/http/HttpClient.h"
+#include "util/http/HttpUtils.h"
+
+// ____________________________________________________________________________
+Service::Service(QueryExecutionContext* qec,
+                 parsedQuery::Service parsedServiceClause)
+    : Operation(qec) {
+  parsedServiceClause_ = parsedServiceClause;
+  using Url = ad_utility::httpUtils::Url;
+
+  // TODO: The following code sends a SPARQL query to the remote endpoint and
+  // turns the TSV result into a `parsedQuery::Values` object. Is this
+  // constructor the right place for this code or should this be in
+  // `computeResult` or somewhere else?
+
+  // Get the URL of the SPARQL endpoint (we can only do fixed IRIs at the
+  // moment).
+  std::string_view serviceIriString = parsedServiceClause_.serviceVarOrIri_;
+  if (serviceIriString.starts_with("?")) {
+    throw std::runtime_error(
+        "SERVICE with variable for endpoint not yet supported by QLever");
+  }
+  AD_CHECK(serviceIriString.starts_with("<") &&
+           serviceIriString.ends_with(">"));
+  serviceIriString.remove_prefix(1);
+  serviceIriString.remove_suffix(1);
+  Url serviceUrl(serviceIriString);
+
+  // Construct the query to be sent to the SPARQL endpoint.
+  std::string serviceQuery =
+      absl::StrCat(parsedServiceClause_.servicePrologue_, "\nSELECT * WHERE ",
+                   parsedServiceClause_.serviceQueryBody_);
+  LOG(INFO) << "Sending SERVICE query to remote endpoint "
+            << "(protocol: " << serviceUrl.protocolAsString()
+            << ", host: " << serviceUrl.host()
+            << ", port: " << serviceUrl.port()
+            << ", target: " << serviceUrl.target() << ")" << std::endl
+            << serviceQuery << std::endl;
+
+  // Send the query and get the result.
+  std::istringstream tsvResult = [&]() {
+    if (serviceUrl.protocol() == Url::Protocol::HTTP) {
+      HttpClient client(serviceUrl.host(), serviceUrl.port());
+      return client.sendRequest(boost::beast::http::verb::post,
+                                serviceUrl.host(), serviceUrl.target(),
+                                serviceQuery, "application/sparql-query",
+                                "text/tab-separated-values");
+    } else {
+      HttpsClient client(serviceUrl.host(), serviceUrl.port());
+      return client.sendRequest(boost::beast::http::verb::post,
+                                serviceUrl.host(), serviceUrl.target(),
+                                serviceQuery, "application/sparql-query",
+                                "text/tab-separated-values");
+    }
```

Comment:
Remove this code duplication by a lambda:
```
 auto sendRequest = [&]<typename Client>() { Client client{host, port}; ...};
```
Usage:
```
if (protocol == HTTP) {
  tsvResult = sendRequest.operator()<HttpClient>();
}
```
(There is an alternative way to implement this that makes the call something like)
`tsvResult = sendRequest(std::type_identity<HttpClient>{})`,
tell me what you prefer:)